### PR TITLE
weighted Katz

### LINKED ISSUE #1423: Katz centrality for weighted graphs
Why in https://github.com/JuliaGraphs/LightGraphs.jl/blob/master/src/Centrality/katz.jl#L46 there is the restriction to `Bool` type. I think that allowing `Float64` should be enough and it would allow for weighted graphs.

### PR DESCRIPTION
**Pull request by sbromberger**
Closes #1423 

cc @bkamins for a quick sanity check, please.

### CODE REVIEW

File: src/Centrality/katz.jl [hunk 1]
```diff
@@ -29,22 +29,31 @@ A struct describing an algorithm to calculate the [Katz centrality](https://en.w
 of the graph `g` optionally parameterized by `α`. Return a vector representing
 the centrality calculated for each node in `g`.
 
+It is also possible to pass in a matrix of weights for weighted Katz centrality.
+
 ### Optional Parameters
 - `α::Real=0.3`: the Katz centrality attenuation factor.
+- `normalize=true`: If true, normalize the betweenness values by scaling the values by `norm`.
```
Comment (bkamins) [inline src/Centrality/katz.jl:36]: probably Katz not betweenness - right?

File: src/Centrality/katz.jl [hunk 2]
```diff
@@ -29,22 +29,31 @@ A struct describing an algorithm to calculate the [Katz centrality](https://en.w
...
-    v /=  norm(v)
+    v = (spI - α * A) \ v
+    if normalize
+        v ./=  norm(v)
+    end
     return v
 end
+
+centrality(g::AbstractGraph, alg::Katz) = _katz_centrality(g, adjacency_matrix(g, Bool; dir=:in), alg.α, alg.normalize)
+centrality(g::AbstractGraph, distmx::AbstractMatrix, alg::Katz) = _katz_centrality(g, transpose(adjacency_matrix(g) .* distmx), alg.α, alg.normalize)
```
Comment (bkamins) [inline src/Centrality/katz.jl:58]: all looks good here from the mathematical perspective. One question is that I see that you allow `distmx` to have non-zero entries even if there are no edges - right?

Also now I understand the problem with transposing you had. What you have implemented looks OK.

File: src/Centrality/katz.jl [hunk 3]
```diff
@@ -29,22 +29,31 @@ A struct describing an algorithm to calculate the [Katz centrality](https://en.w
...
-    v = (spI - alg.α * A) \ v
-    v /=  norm(v)
+    v = (spI - α * A) \ v
+    if normalize
+        v ./=  norm(v)
+    end
     return v
 end
+
+centrality(g::AbstractGraph, alg::Katz) = _katz_centrality(g, adjacency_matrix(g, Bool; dir=:in), alg.α, alg.normalize)
```
Comment (bkamins) [inline src/Centrality/katz.jl:57]: this method is not tested for weighted graphs, but I guess it cannot be helped - right?

File: src/Centrality/katz.jl (hunk 1 above)
Comment (sbromberger) [reply src/Centrality/katz.jl]: Oops. Fixed.

File: src/Centrality/katz.jl (hunk 2 above)
Comment (sbromberger) [reply src/Centrality/katz.jl:58]: yes, that's a deliberate decision about distance matrices: the values are only defined at valid [i, j] pairs - using anything else is undefined behavior.

This allows us to make a `DefaultDistance <: AbstractMatrix` which equals `1` for any `getindex` and have it be O(1).

File: src/Centrality/katz.jl (hunk 3 above)
Comment (sbromberger) [reply src/Centrality/katz.jl:57]: `SimpleWeightedGraphs` is the proper place to override this. We should do that post-merge.

**Comment by sbromberger** [comment]
I plan on merging this at the end of the day tomorrow, Friday 22 May unless there are objections.

@bkamins - I don't think I'm going to port this to 1.3.x - are you ok with working from master for a bit?

**Comment by bkamins** [comment]
I will work on master anyway :). Thank you for caring so much.

**Pull request merged**